src/tools/rest-backend/root_application.cpp

```diff
@@ -0,0 +1,308 @@
+#include <boost/algorithm/string/replace.hpp>
+#include <cppcms/view.h>
+#include <sstream>
+
+#include "authentication_application.hpp"
+#include "conversion_application.hpp"
+#include "database_application.hpp"
+#include "exceptions.hpp"
+#include "kdb.h"
+#include "root_application.hpp"
+#include "service.hpp"
+#include "user_application.hpp"
+
+
+namespace kdbrest
+{
+
+/**
+     * @brief constructor for main service class
+     *
+     * @param srv a service container
+     */
+RootApp::RootApp (cppcms::service & srv) : cppcms::application (srv)
+{
+
+	attach (new AuthenticationApp (srv), "auth", "/auth{1}", // mapping
+		"/auth(/(.*))?", 1				 // dispatching
+		);
+
+	attach (new UserApp (srv), "user", "/user{1}", // mapping
+		"/user(/(.*))?", 1		       // dispatching
+		);
+
+	attach (new DatabaseApp (srv), "database", "/database{1}", // mapping
+		"/database(/(.*))?", 1				   // dispatching
+		);
+
+	attach (new ConversionApp (srv), "conversion", "/conversion{1}", // mapping
+		"/conversion(/(.*))?", 1				 // dispatching
+		);
+
+	dispatcher ().assign ("", &RootApp::welcome, this);
+	mapper ().assign ("");
+
+	dispatcher ().assign ("/version", &RootApp::version, this);
+	mapper ().assign ("version", "/version");
+
+	mapper ().root ("");
+}
+
+/**
+ * @brief handler for the root of the whole rest service, returns an API description.
+ */
+void RootApp::welcome ()
+{
+	RootApp::setCORSHeaders (response (), "GET,OPTIONS");
+
+	if (request ().request_method () == "GET")
+	{
+		const std::string raw = request ().get (PARAM_RAW);
+		if (!raw.empty () && raw == "true")
+		{
+			RootApp::setSeeOther (response (), cppcms::application::settings ().get ("api_specification.raw", ""));
+			return;
+		}
+
+		RootApp::setSeeOther (response (), cppcms::application::settings ().get ("api_specification.html", ""));
+		return;
+	}
+	else if (request ().request_method () == "OPTIONS")
+	{
+		RootApp::setOk (response ());
+		return;
+	}
+	else
+	{
+		RootApp::setMethodNotAllowed (response ()); // send HTTP 405
+		return;
+	}
+}
+
+/**
+ * @brief handler for the version resource of the endpoint, returns versions for the API
+ *		  and the used Elektra version.
+ */
+void RootApp::version ()
+{
+	RootApp::setCORSHeaders (response (), "GET,OPTIONS");
+
+	if (request ().request_method () == "GET")
+	{
+		cppcms::json::value data;
+
+		data["elektra"]["version"] = KDB_VERSION;
```

Comment:
This is the Elektra version build against, for the run-time values you need to query `system/elektra/version/constants`